CI: add attributes check run

kamilrakoczy opened the pull request:
Related to: https://github.com/verilator/verilator/pull/3608

This PR adds ``clang_check_attributes.py`` that is using clang python bindings to make sure that if function is marked as ``VL_MT_SAFE`` it calls only functions that are also marked as ``VL_MT_SAFE``. For now it doesn't fail the CI run, but when all functions will satisfy this condition, it can be changed.

This CI run checks for 2 things:
 * if function is marked as ``VL_MT_SAFE``, it only calls other functions marked as ``VL_MT_SAFE``, if it is not the case, it prints tree of calls that leads to this unsafe call

 * if function is marked as ``VL_MT_START``, it prints all function calls to help finding out what functions are called
 
 I'm open to suggestions, I'm especially not sure if we want to print whole callstack of ``VL_MT_SAFE`` in the CI, but this feature is very helpful when developer wants to add new function that will run in parallel context.
 
 Before merge I will remove 6955e4dc074d9f86a9dcb2a42c0c21660e9ab0a3 commit, It was only added to showcase how ``VL_MT_START`` feature will look like.

Signed-off-by: Kamil Rakoczy <krakoczy@antmicro.com>

wsnyder commented on .github/workflows/build.yml:
```diff
@@ -158,3 +158,39 @@ jobs:
+
+  attributes-check:
+    needs: build
+    runs-on: ubuntu-20.04
+    name: Ubuntu 20.04 | Attributes check
```
Please use 22.04 as it will last longer and that is where I do development.

However can we instead just run this as a test_regress/t/t_dist_attributes.pl as part of normal regressions (when on 22.04)?   That saves a whole checkout and works better for my personal flow.

kamilrakoczy replied:
> Please use 22.04 as it will last longer and that is where I do development.

At first, I've used 22.04, but it turned out that verilator is only build for 20.04 in PRs.
I need source code after build as some parts of the files are generated using ``astgen``.
 
> However can we instead just run this as a test_regress/t/t_dist_attributes.pl as part of normal regressions (when on 22.04)? That saves a whole checkout and works better for my personal flow.

Sure, I will try to do this.

kamilrakoczy replied:
I switched to ``test_regress`` format, but I'm unsure how to print test output in the CI without failing the test and re-running it. Test takes about 6-7 minutes to check all files and it seems like waste of CI time to re-run it.

There are some unsafe calls in verilator, so for now, I decided to create golden file and compare result to the golden file, but in the future we probably would like to fail test when we detect any unsafe call.

It should be ready for review.

wsnyder commented on ci/clang_check_attributes.py:
```diff
@@ -0,0 +1,285 @@
+#!/usr/bin/env python3
+#
+# Copyright 2022 by Antmicro. Verilator is free software; you
```
I'd prefer it be copyright under my name as the rest of the package, but if you all feel strongly this can work - but then change "Verilator is" to "This is"

kamilrakoczy replied:
Sure, I've changed copyright to your name.

wsnyder commented on ci/clang_check_attributes.py:
```diff
@@ -0,0 +1,285 @@
+#!/usr/bin/env python3
```
I think this is better located under "nodist" as it is usable outside the ci tool.

wsnyder commented on ci/clang_check_attributes.py:
```diff
@@ -0,0 +1,285 @@
+def print_node(node, level):
+    if fully_qualified_pretty(node) not in PRINTED:
+        annotations = get_annotations(node)
+        if "MT_SAFE" in annotations or node.kind in MT_SAFE_KINDS:
```
Please also look for VL_PURE.  Pure means a function has no external dependencies and is thus also VL_MT_SAFE.

Please check a VL_PURE function can only call another VL_PURE function not one that is VL_MT_SAFE.

wsnyder reviewed (commented):
Great!

If you didn't already please make sure this code is checked (after moving to nodist) by the "make lint-py" rule and is completely clean by that check.

wsnyder commented:
P.S. Please also update from master, as I fixed a few VL_PURE markers.

wsnyder commented:
We can move PRs to 22.04, feel free to make another pull.

wsnyder commented on Makefile.in:
```diff
@@ -379,6 +379,7 @@ PY_PROGRAMS = \
 	nodist/fuzzer/actual_fail \
 	nodist/fuzzer/generate_dictionary \
 	nodist/install_test \
+	nodist/clang_check_attributes.py \
```
Move up to keep sorted please.  Also I would suggest drop .py extension, as it's an executable and the language doesn't really matter.

wsnyder commented on ci/ci-install.bash:
```diff
@@ -93,6 +93,18 @@ elif [ "$CI_BUILD_STAGE_NAME" = "test" ]; then
     # libfl-dev needed for internal coverage's test runs
     sudo apt-get install gdb gtkwave lcov libfl-dev ccache
+    # Required for test_regress/t/t_dist_attributes.pl
+    sudo apt-get install libclang-dev
+    if [ "$CI_RUNS_ON" = "ubuntu-22.04" ]; then
+      pip3 install clang==14.0
+    fi
+    if [ "$CI_RUNS_ON" = "ubuntu-20.04" ]; then
+      pip3 install clang==10.0.1
+    fi
+    if [ "$CI_RUNS_ON" = "ubuntu-18.04" ]; then
+      sudo apt-get install python3-setuptools
+      pip3 install clang==6.0.0.2
+    fi
```
```suggestion
    if [ "$CI_RUNS_ON" = "ubuntu-22.04" ]; then
      sudo apt-get install libclang-dev
      pip3 install clang==14.0
    fi
```

I suggest to keep our support easier we only run this test on 22.04.  Then in the t_.pl file, check the OS version and do a skip(....) if it isn't 22.04.  This will also resolve some of the differences you mentioned in reports between versions.

wsnyder commented on test_regress/t/t_dist_attributes.pl:
```suggestion
    run(logfile => $self->{run_log_filename},
        tee => 1,
        cmd => ["python3", "$root/nodist/clang_check_attributes.py $input_dirs $include_dirs -x $exclude"]);

    file_grep($self->{run_log_filename}, "Number of functions marked as MT_SAFE calling unsafe functions: 72");
```

wsnyder commented on nodist/clang_check_attributes.py:
```diff
@@ -0,0 +1,325 @@
+                self.pop(next(iter(self)))
+
+
+INDEX = Index.create()
+CLANG_ARGS = []
+TRANSLATE_UNITS = FixSizedDict(maxlen=20)
+PRINTED = []
+
+# by definition constructor is MT safe,
+# but functions that constructor call are not!
+MT_SAFE_KINDS = [CursorKind.CONSTRUCTOR]
+UNSAFE_CALLS = []
```
All caps for true constants is ok (e.g. VERILATOR_ROOT), but please use lower case with underscores for global variables as that is the Python way https://stackoverflow.com/questions/159720/what-is-the-naming-convention-in-python-for-variable-and-function

wsnyder commented on nodist/clang_check_attributes.py:
```diff
@@ -0,0 +1,325 @@
+    except KeyError:
+        translate_unit = INDEX.parse(file, cfg['clang_args'])
+        if not translate_unit:
+            print(f"unable to load translate_unit from file: {file}")
```
```suggestion
            print(f"%Error: unable to load translate_unit from file: {file}")
```

Or if the % bothers you, just Error:

wsnyder commented on nodist/clang_check_attributes.py:
```diff
@@ -0,0 +1,325 @@
+def get_files(filename):
+    files = list(glob.glob(f"{filename}/*.cpp"))
+    # Dont process test files
+    files = [f for f in files if "_test.cpp" not in f]
+    # Dont process Verilator coverage files
+    files = [f for f in files if "Vlc" not in f]
+    # We want to use V3Const after astgen
+    files = [f for f in files if "V3Const.cpp" not in f]
+    # Skip SystemC files
+    files = [f for f in files if "_sc" not in f]
+    # Skip parser files
+    files = [f for f in files if ".yy.cpp" not in f]
```
Perhaps cleaner to have a list of regexps to ignore at top of program, then iterate that here?

wsnyder commented on nodist/clang_check_attributes.py:
```diff
@@ -0,0 +1,325 @@
+    excluded_prefixes = []
+    excluded_paths = ['/usr']
+    i = 0
+    while i < len(args):
+        if args[i] == '-x':
+            i += 1
+            excluded_prefixes = args[i].split(',')
+        elif args[i] == '-p':
+            i += 1
+            excluded_paths = args[i].split(',')
+        elif args[i][0] == '-':
+            clang_args.append(args[i])
+        else:
+            base_folders.append(args[i])
+        i += 1
```
Please use https://docs.python.org/3/library/argparse.html
for examples see the other python programs in Verilator, e.g. src/arggen.

wsnyder commented on nodist/clang_check_attributes.py:
```diff
@@ -0,0 +1,325 @@
+
+
+def main():
+    if len(sys.argv) < 2:
+        print('usage: ' + sys.argv[0] + ' folder [extra clang args...]')
+        return
```
Remove as argparse will do it for you.

wsnyder commented on ci/ci-install.bash:
```diff
@@ -93,6 +93,18 @@ elif [ "$CI_BUILD_STAGE_NAME" = "test" ]; then
     # libfl-dev needed for internal coverage's test runs
     sudo apt-get install gdb gtkwave lcov libfl-dev ccache
+    # Required for test_regress/t/t_dist_attributes.pl
+    sudo apt-get install libclang-dev
```
Add that libclang-dev is recommended for developers to docs/guide/install.rst.

wsnyder commented on test_regress/t/t_dist_attributes.pl:
Let's make a t_dist_attributes_src.pl and t_dist_attributes_include.pl which run on src/ and include/ separately.  Then they can run in parallel.  (A few include files are used by both so might cause warnings on both, but that's a very small set.)

The total runtime might still be the long path in my personal regressions, maybe I'll add a way to skip it with an environment variable.

wsnyder commented on nodist/clang_check_attributes.py:
```diff
@@ -0,0 +1,325 @@
+        filepath = pathlib.Path(str(node.location.file)).resolve()
+        try:
+            relative = filepath.relative_to(VERILATOR_ROOT)
+            print('%s %-35s %-100s %s [%s:%s - %s:%s] \t %s  %s' %
+                  (' ' * level, node.kind, fully_qualified_pretty(node), ' ' *
+                   (20 - level), node.extent.start.line,
+                   node.extent.start.column, node.extent.end.line,
+                   node.extent.end.column, relative, annotations))
```
I suggest print the filename and line number with colon first on the line (e.g. src/V3Foo.h:1888: ... ) and omit the end line/column info, so most editors can jump there with a single key press.

kamilrakoczy replied:
Sure, now I'm printing:
```
 include/verilated_vcd_c.h:281           CursorKind.CXX_METHOD               VerilatedVcdC::flush()                                                                               ['MT_SAFE']
  include/verilated_vcd_c.cpp:270                CursorKind.CXX_METHOD               VerilatedVcd::flush()                                                                                []
   include/verilated.h:189               CursorKind.CONSTRUCTOR              VerilatedLockGuard::VerilatedLockGuard(VerilatedMutex &)                                             ['MT_SAFE']
    include/verilated.h:163              CursorKind.CXX_METHOD               VerilatedMutex::lock()                                                                               ['MT_SAFE']
     include/verilated.h:177             CursorKind.CXX_METHOD               VerilatedMutex::try_lock()                                                                           ['MT_SAFE']
```

wsnyder commented on test_regress/t/t_dist_attributes_src.pl:
```suggestion
my $clang_args = "-I$root/src/ -I$root/include/ -I$root/src/obj_opt/ -fcoroutines-ts";
```

For source run vltstd not needed.

wsnyder commented on test_regress/t/t_dist_attributes_include.pl:
```suggestion
        tee => 1,
        cmd => ["python3", "$root/nodist/clang_check_attributes $input_dirs $clang_args -x $exclude"]);
```

Below too.  There's probably a way to use "do" to put the common code in a single file but I can live with this as-is.

wsnyder commented on test_regress/t/t_dist_attributes_include.pl:
Please add an 

if ($ENV{VERILATOR_TEST_NO_ATTRIBUTES}) {
    skip("Skipping due to VERILATOR_TEST_NO_ATTRIBUTES");
} else {
    check();
}

and move the other code here to "sub check".

wsnyder commented on nodist/clang_check_attributes:
```diff
@@ -0,0 +1,332 @@
+        'ranges': list(diag.ranges),
+        'fixits': list(diag.fixits)
+    }
+
```
```suggestion
```

Please put single line between functions, only double line is where "make format" forces them.

kamilrakoczy replied:
``make format`` forces double line between functions.

wsnyder commented on nodist/clang_check_attributes:
```diff
@@ -0,0 +1,332 @@
+        filepath = Path(str(node.location.file)).resolve()
+        try:
+            filepath_relative = filepath.relative_to(VERILATOR_ROOT)
+            print('%s %s:%s \t\t %-35s %-100s %s' %
+                  (' ' * level, filepath_relative, node.extent.start.line,
+                   node.kind, fully_qualified_pretty(node), annotations))
```
To appease editors, the filename needs to be on left, with colon after.  Also I don't think CursorKind adds useful information, from reading the method name it's pretty obvious (if you think is critical for debug then keep it but remove "CursorKind." prefix).  Also having attributes on far right adds a lot of spacing and generally IMO looks better to have all fixed indents on left.  Finally please don't use tabs.  So here's the first output I'd expect:

```
::group::VerilatedContext::commandArgs(int, char **)
include/verilated.h:418:             ['MT_SAFE']    VerilatedContext::commandArgs(int, char **)
include/verilated.cpp:2494:          []              VerilatedContext::commandArgs(int, const char **)
include/verilated.h:189:             ['MT_SAFE']      VerilatedLockGuard::VerilatedLockGuard(VerilatedMutex &)
include/verilated.h:163:             ['MT_SAFE']       VerilatedMutex::lock()
include/verilated.h:177:             ['MT_SAFE']        VerilatedMutex::try_lock()
include/verilated.h:536:             []               VerilatedContext::impp()
include/verilated_imp.h:390:         []               VerilatedContextImp::commandArgsAddGuts(int, const char **)
::endgroup::
```

wsnyder reviewed (commented):
The include test takes only 17 seconds which is great.  Wish src was that speed!

wsnyder commented:
Also:

    include/verilated.cpp:2494             CursorKind.CXX_METHOD               VerilatedContext::commandArgs(int, const char **)                                                    []
safe due to VL_MT_SAFE_EXCLUDES(m_argMutex);

    include/verilated_imp.h:390            CursorKind.CXX_METHOD               VerilatedContextImp::commandArgsAddGuts(int, const char **)                                          []
safe due to VL_REQUIRES(m_argMutex)

    include/verilated_imp.h:435            CursorKind.CONSTRUCTOR              VerilatedImpData::VerilatedImpData()                                                                 []
C++ guarantees static constructors are safe.  Not sure how we handle this, perhaps mark safe anyways?

    include/verilated_imp.h:266            CursorKind.CXX_METHOD               VerilatedContextImp::argPlusMatch(const char *)                                                      []
safe due to VL_MT_SAFE_EXCLUDES(m_argMutex) {

```
   ::group::VerilatedCovImp::keyValueFormatter(const std::string &, const std::string &)
    include/verilated_cov.cpp:165           CursorKind.CXX_METHOD               VerilatedCovImp::keyValueFormatter(const std::string &, const std::string &)                         ['PURE']
     include/verilated_cov.cpp:141                  CursorKind.CXX_METHOD               VerilatedCovImp::dequote(const std::string &)                                                        ['PURE']
     ::endgroup::
This and similar ones shouldn
```'t be printed, PURE under PURE is ok.

wsnyder commented:
I pushed a few include cleanups.

I think you should also add a t_dist_attributes_bad.pl and .cpp test that checks that the program properly reports and suppresses failures, this will make it a lot easier to debug and maintain.

kamilrakoczy commented:
> Also:
> 
> ```
> include/verilated.cpp:2494             CursorKind.CXX_METHOD               VerilatedContext::commandArgs(int, const char **)                                                    []
> ```
> 
> safe due to VL_MT_SAFE_EXCLUDES(m_argMutex);

Fixed

> 
> ```
> include/verilated_imp.h:390            CursorKind.CXX_METHOD               VerilatedContextImp::commandArgsAddGuts(int, const char **)                                          []
> ```
> 
> safe due to VL_REQUIRES(m_argMutex)

Fixed

> 
> ```
> include/verilated_imp.h:435            CursorKind.CONSTRUCTOR              VerilatedImpData::VerilatedImpData()                                                                 []
> ```
> 
> C++ guarantees static constructors are safe. Not sure how we handle this, perhaps mark safe anyways?

I've already marked constructors as MT_SAFE, but I didn't print it in attributes. Now I added ``MT_SAFE_DEFINITION`` to annotations for constructors.

> 
> ```
> include/verilated_imp.h:266            CursorKind.CXX_METHOD               VerilatedContextImp::argPlusMatch(const char *)                                                      []
> ```
> 
> safe due to VL_MT_SAFE_EXCLUDES(m_argMutex) {

Fixed

> 
> ```
>    ::group::VerilatedCovImp::keyValueFormatter(const std::string &, const std::string &)
>     include/verilated_cov.cpp:165           CursorKind.CXX_METHOD               VerilatedCovImp::keyValueFormatter(const std::string &, const std::string &)                         ['PURE']
>      include/verilated_cov.cpp:141                  CursorKind.CXX_METHOD               VerilatedCovImp::dequote(const std::string &)                                                        ['PURE']
>      ::endgroup::```
> 
> This and similar ones shouldn't be printed, PURE under PURE is ok.
> ```

Actually it was another problem, I didn't print nodes that are defined outside verilator repository, but I checked if they have annotation. I fixed this by adding extra parameter that allows to skip functions defined in specific folders, in this case in ``/usr``.


> I think you should also add a t_dist_attributes_bad.pl and .cpp test that checks that the program properly reports and suppresses failures, this will make it a lot easier to debug and maintain.

I agree and I started working on this test, but if you agree, maybe lets leave it to another PR to do not stale this one?

wsnyder commented:
I made a mistake above; VL_REQUIRES means a caller had to get the lock so it is itself VL_MT_UNSAFE. If a VL_MT_SAFE calls a function that is VL_REQUIRES it is missing a lock and so should be flagged.

Also I realize I think you need a mark to know when a function has a mutex so everything underneath it is safe, let's label those VL_MT_ACQUIRES. Any use of VL_ACQUIRE(x) can also be marked VL_MT_ACQUIRES.

1. A function with VL_PURE can only call another function that is VL_PURE.

2. A function with VL_MT_SAFE can only call a function that is VL_PURE, VL_MT_SAFE, or VL_MT_ACQUIRES. Once a function with VL_MT_ACQUIRES is hit in it's child calls, any further functions below that VL_MT_ACQUIRES one is no longer processed (it's ok as has lock).

kamilrakoczy commented:
> I made a mistake above; VL_REQUIRES means a caller had to get the lock so it is itself VL_MT_UNSAFE. If a VL_MT_SAFE calls a function that is VL_REQUIRES it is missing a lock and so should be flagged.
> 
> Also I realize I think you need a mark to know when a function has a mutex so everything underneath it is safe, let's label those VL_MT_ACQUIRES. Any use of VL_ACQUIRE(x) can also be marked VL_MT_ACQUIRES.
> 
> 1. A function with VL_PURE can only call another function that is VL_PURE.
> 2. A function with VL_MT_SAFE can only call a function that is VL_PURE, VL_MT_SAFE, or VL_MT_ACQUIRES. Once a function with VL_MT_ACQUIRES is hit in it's child calls, any further functions below that VL_MT_ACQUIRES one is no longer processed (it's ok as has lock).

Sure, I will add this functionality. For now I reverted changes that adds ``VL_MT_SAFEREQUIRES``.

I've also added ``t_dist_attributes_bad.pl`` test.

wsnyder commented:
```
::group::VerilatedContext::commandArgs(int, char **)
include/verilated.h:418:                           ['MT_SAFE']                               VerilatedContext::commandArgs(int, char **)
include/verilated.cpp:2456:                        ['MT_SAFE']                                VerilatedContext::commandArgs(int, const char **)
include/verilated.h:536:                           ['MT_SAFE']                                 VerilatedContext::impp()
include/verilated.h:189:                           ['MT_SAFE', 'MT_SAFE_DEFINITION']           VerilatedLockGuard::VerilatedLockGuard(VerilatedMutex &)
include/verilated.h:163:                           ['MT_SAFE']                                  VerilatedMutex::lock()
include/verilated.h:177:                           ['MT_SAFE']                                   VerilatedMutex::try_lock()
include/verilated_imp.h:390:                       []                                          VerilatedContextImp::commandArgsAddGuts(int, const char **)
```

This is great stuff, it's finding a lot of missing annotations. 

Can we combine MT_SAFE_DEFINITION to not print MT_SAFE or otherwise to make the middle column less wide?

VerilatedContext::commandArgs gets a mutex so everything after is fine.  Should we be marking this VL_MT_ACQUIRES?  I don't see that in what you supported.  I think MT_SAFE_EXCLUDES could perhaps serve that purpose, we then would allow anything under an EXCLUDES. Ideally we should check that any EXCLUDES function has itself a call to VerilatedMutex.

Could we pass all the VL_MT_* flavors into the tool so we can see the exact labeling in the report?

Can you recognize constexpr functions and automatically mark them pure (see activeMagic() which for now I did mark pure, but that seems redundant if you can tell the constexpr).

Please merge in master again, I made more cleanups based on the reports.  If I did it right, with the above EXCLUDES it should be nearly clean for include.  src we still need to merge your #3729.

wsnyder commented:
Please rename t_dist_attributes* to t_a5_attributes*, so that these long running tests start almost first, thanks.

wsnyder commented:
Please merge from master and add to the t_*.pl files:

rerunnable(0);

this will stop them from rerunning if they fail.

mglb commented:
> I made a mistake above; VL_REQUIRES means a caller had to get the lock so it is itself VL_MT_UNSAFE. If a VL_MT_SAFE calls a function that is VL_REQUIRES it is missing a lock and so should be flagged.
> (...)
>     2. A function with VL_MT_SAFE can only call a function that is VL_PURE, VL_MT_SAFE, or VL_MT_ACQUIRES. **Once a function with VL_MT_ACQUIRES is hit in it's child calls, any further functions below that VL_MT_ACQUIRES one is no longer processed (it's ok as has lock).**

Isn't such a check superfluous? One of the main purposes of Clang's `-Wthread-safety` is checking whether a function with REQUIRES(mtx) is called after a function marked as ACQUIRE(mtx), but before a function marked as RELEASE(mtx). This also checks whether the correct mutex is locked. Example: https://clang.llvm.org/docs/ThreadSafetyAnalysis.html#getting-started

I think we should just assume that functions with VL_REQUIRES, and any method called on a member with `VL_GUARDED_BY`, is VL_MT_SAFE. Clang will check the rest.

wsnyder commented:
>I think we should just assume that functions with VL_REQUIRES, and any method called on a member with VL_GUARDED_BY, is VL_MT_SAFE. 

Yes, but we also need to not look below those functions, as we would with a normal VL_MT_SAFE.  Once a lock is present anything it calls is presumably safe as won't be multithreaded.

mglb commented:
Hi. I continued the work on Kamil's script and it basically turned into a new script at the end.
I took into account probably everything commented here so far + there are a few positive changes.

## Changes

- Every translation unit is analyzed independently (see below).
- Support for parallel parsing and analysis.
- Support for header precompilation (speeds up parsing when done for commonly included headers)
	- Just use `-c HEADER_PATH`, like `-c ./src/V3Ast.h`.
- Faster execution. Execution times with `V3Ast.h` precompilation and analysis of all relevant `.cpp` files in `src/` + `src/obj_opt/V3Const__gen.cpp`, on my machine (6 cores/12 threads CPU + SSD):
	- 12 jobs: 0:52
	- 6 jobs 0:58
	- 2 jobs: 2:38
	- 1 job: 5:05
- Added support for a few new rules (see below).
- Modified defines with attributes in `verilated.h` (such as `VL_ACQUIRE`) to also emit an annotation - most attributes are not exposed to Clang Python API.
- Extra (optional) output formats (mostly for my convenience, can be removed from the PR if needed):
	- `--output-style=bottom-up`: output organized by called function, each with a list of call locations.
	- `--gh-groups=(yes|no|auto)` whether to use grouping syntax used by Github. `auto` (default) means `yes` when `GITHUB_ACTIONS` environment variable is present.
	- `--terminal-format=(yes|no|auto)` uses styles such as bold, italic, etc in the output. `auto` (default) means `yes` when stdout is a terminal.

## Annotation checking

The scripts analyzes every `.cpp` file (AKA compilation unit) independently. Headers included by the file are also processed as a part of the compilation unit. This follows the way how C++ compilers are used.
Due to this independent analysis, annotations must be attached to **declarations** of functions/methods, in header files. Additionally they can also be attached to definitions, but it is not required.
The script emits a warning when annotations on a declaration are not superset of annotations on its definition.

Functions and methods are not checked recursively. The script analyzes direct calls inside all non-excluded definitions (AKA "context").

Excluded are:
- everything from headers outside of the Verilator root
- everything inside `std` namespace
- all symbols with `__` prefix

### Rules

"mt-safe" context is inside definitions of:
- Functions and methods with at least one of the following annotations:
	- `VL_MT_SAFE`
	- `VL_MT_START`
	- `VL_MT_SAFE_POSTINIT`
- ...and without all of the following annotations:
	- `VL_MT_UNSAFE`
	- `VL_MT_UNSAFE_ONE`

"mt-safe" calls:
- Don't have any of the following annotations:
	- `VL_MT_UNSAFE`
	- `VL_MT_UNSAFE_ONE`
  And meet at least one of the following conditions:
	- Functions, function pointers, and methods with at least one of the following annotations:
		- `VL_MT_SAFE`
		- `VL_MT_SAFE_POSTINIT`
		- `VL_PURE`
		- `VL_REQUIRES(...)`
		- `VL_EXCLUDES(...)`, `VL_MT_SAFE_EXCLUDES(...)`
		- `VL_ACQUIRE(...)`, `VL_ACQUIRE_SHARED(...)`
		- `VL_RELEASE(...)`, `VL_RELEASE_SHARED(...)`
	- Constructors
	- All methods called on the following types of objects:
		- locals
		- calling function/method arguments
		- annotated with `VL_GUARDED_BY(...)`

"pure" context is inside definitions of:
- Functions and methods with `VL_PURE` annotation.

"pure" calls:
- Functions and methods with `VL_PURE` annotation.

wsnyder commented:
I'm glad to have it faster!

* Please merge in master, I fixed some python warnings, and fixed some more markers.
* Please make sure it is "make lint-py" clean, by fixing or by adding warning disables.  FIrst update these additional disables at the top:

      # pylint: disable=C0114,C0115,C0116,C0209,C0302,R0902,R0911,R0912,R0914,R0915

* I expect to mostly use this with the test_regress wrappers, so for those wrappers
** Let's use -j 3 (as github has only 2 CPUs)
** Enable header precompilation,

* `--output-style=bottom-up`: I suggest decide what format looks best and just support that
* `--gh-groups=(yes|no|auto)`: I suggest hardcode to no and remove option, as once we get the report clean it will be empty in github (see below example), and th$
* `--terminal-format=(yes|no|auto)`: I prefer to just have non-termal output and remove rest, as otherwise painful to cut-paste
* I prefer we annotate both declaration and definitions, thanks for checking for consistency.

Running it:

First couple of lines I don't understand what it is saying, maybe debug that needs suppression?

    ../include/gtkwave/fstapi.c:5012: value_change_callback    CursorKind.PARM_DECL
        from: ../include/gtkwave/fstapi.c:5254

I think we should start each message "context report" with clean error message describing what seems wrong

```
%Error: 'VL_CVT_PACK_STR_NQ(QData)' is pure but calls non-pure function(s)
include/verilated_funcs.h:2173:     [pure]         VL_CVT_PACK_STR_NQ(QData)
include/verilated_funcs.h:2172:     [mt_safe]        VL_CVT_PACK_STR_NW(int, const WDataInP)
include/verilated_types.h:203:      []               VlWide<2>::VlWide()
include/verilated_types.h:215:      []               VlWide<2>::operator unsigned int *()

%Error: 'VL_FCLOSE_I(IData)' declaration does not match definition
include/verilated_funcs.h:122:      []             VL_FCLOSE_I(IData) [declaration]
include/verilated.cpp:1426:         [mt_safe]      VL_FCLOSE_I(IData)
```

How do we prevent this one getting flagged?

     include/verilated_types.h:203:      []               VlWide<2>::VlWide()

This is an example where we have a static lock, we need a way to mark this so the addCb is not flagged, VL_MT_SAFE_GUARDED?

    include/verilated.cpp:2791:        [mt_safe]    Verilated::addFlushCb(Verilated::VoidPCb, void *)
    include/verilated.cpp:2778:        []             addCb(Verilated::VoidPCb, void *, VoidPCbList &)

kamilrakoczy commented:
> I'm glad to have it faster!
> 
> * Please merge in master, I fixed some python warnings, and fixed some more markers.
> * Please make sure it is "make lint-py" clean, by fixing or by adding warning disables.  FIrst update these additional disables at the top:
>   ```
>   # pylint: disable=C0114,C0115,C0116,C0209,C0302,R0902,R0911,R0912,R0914,R0915
>   ```

Done, I've also added ``E1101`` due to false-positive report on missing member in ``Field``:
```
Instance of 'Field' has no 'add' member (no-member)
```

> * I expect to mostly use this with the test_regress wrappers, so for those wrappers
> * * Let's use -j 3 (as github has only 2 CPUs)

By default script uses ``max(1, len(os.sched_getaffinity(0)))`` so it was already done.

> * * Enable header precompilation,

We tested that only ``V3Ast.h`` is large enough to significantly reduce time and it is already enabled in test_regress wrapper.

> * `--output-style=bottom-up`: I suggest decide what format looks best and just support that

Sure, I left only ``top-down`` style.

> * `--gh-groups=(yes|no|auto)`: I suggest hardcode to no and remove option, as once we get the report clean it will be empty in github (see below example)

Removed.

> * `--terminal-format=(yes|no|auto)`: I prefer to just have non-termal output and remove rest, as otherwise painful to cut-paste

Removed.

> * I prefer we annotate both declaration and definitions, thanks for checking for consistency.
> 
> Running it:
> 
> First couple of lines I don't understand what it is saying, maybe debug that needs suppression?
> 
> ```
> ../include/gtkwave/fstapi.c:5012: value_change_callback    CursorKind.PARM_DECL
>     from: ../include/gtkwave/fstapi.c:5254
> ```
> 

Fixed.

> I think we should start each message "context report" with clean error message describing what seems wrong
> 
> ```
> %Error: 'VL_CVT_PACK_STR_NQ(QData)' is pure but calls non-pure function(s)
> include/verilated_funcs.h:2173:     [pure]         VL_CVT_PACK_STR_NQ(QData)
> include/verilated_funcs.h:2172:     [mt_safe]        VL_CVT_PACK_STR_NW(int, const WDataInP)
> include/verilated_types.h:203:      []               VlWide<2>::VlWide()
> include/verilated_types.h:215:      []               VlWide<2>::operator unsigned int *()
> 
> %Error: 'VL_FCLOSE_I(IData)' declaration does not match definition
> include/verilated_funcs.h:122:      []             VL_FCLOSE_I(IData) [declaration]
> include/verilated.cpp:1426:         [mt_safe]      VL_FCLOSE_I(IData)
> ```
> 

Sure, changed.

> How do we prevent this one getting flagged?
> 
> ```
>  include/verilated_types.h:203:      []               VlWide<2>::VlWide()
> ```
> 

Fixed, I've added check that marks default constructors as ``VL_PURE``.

> This is an example where we have a static lock, we need a way to mark this so the addCb is not flagged, VL_MT_SAFE_GUARDED?
> 
> ```
> include/verilated.cpp:2791:        [mt_safe]    Verilated::addFlushCb(Verilated::VoidPCb, void *)
> include/verilated.cpp:2778:        []             addCb(Verilated::VoidPCb, void *, VoidPCbList &)
> ```
I think we shouldn't add new ``VL_MT_SAFE_GUARDED``, but instead we should refactor the code and properly mark functions with ``VL_REQUIRES``/``VL_EXCLUDES``. This way clang will warn on possible dead-locks and will prevent using this functions without locking mutex first. I've prepared example refactor in: https://github.com/antmicro/verilator-1/commit/83239ee183f7a9a23ae779be7b987b84b8a97502#diff-1c0f1298e2ed5505c79d957359eb3bdfee1dfa2ca6efdb67c2f9a28126639804L2794


I've also noticed that https://github.com/verilator/verilator/commit/86e3c02339c48b81ce8673e0c4adeb41d0bdd000 introduced some weird CI behavior, before this commit, we didn't delete ``V3Const__gen.cpp`` after building verilator (see: https://github.com/verilator/verilator/actions/runs/3666742393/jobs/6198703968#step:5:464), after this commit, we started to delete this file: https://github.com/verilator/verilator/actions/runs/3669473981/jobs/6203327840#step:5:465.

I don't entire understand why this is happening, but it looks like restoring ``V3__CONCAT.cpp`` rule (https://github.com/verilator/verilator/pull/3748/commits/b6ffeefc36213439366120f4ec302c9d0d71d60b)  preserves this file after building verilator.

wsnyder commented:
>I've also noticed that https://github.com/verilator/verilator/commit/86e3c02339c48b81ce8673e0c4adeb41d0bdd000 introduced some weird CI behavior (at least for me), before this commit, we didn't delete V3Const__gen.cpp

This is fixed.

>I think we shouldn't add new VL_MT_SAFE_GUARDED, but instead we should refactor the code and properly mark functions with VL_REQUIRES/VL_EXCLUDES.

I think it's easier to read keeping the statics, but agree probably stomping out the function-local statics is better code.  I'm not sure of some of your markings, please make a new pull for that and I'll comment on it.  Please first merge master, as I committed some of your VL_PURE markings from there plus added some additional ones.

My thinking is once we have the includes clean, this can check for no include failures, then we can commit the attribute checker.
